File: runner-js/runner.js
#!/usr/bin/env node

var readline = require('readline');
var rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false
});

var type = 'code';
var code = '';
var runner = null;

rl.on('line', function(line){
    if (type === 'code') {
        if (line.startsWith('__END__CODE__')) {
            type = 'run';
            console.log("READY");
        } else {
            code += line;
            code += "\n";
        }
    } else if (type === 'run') {
        if (line.startsWith('EXIT')) {
            process.exit(0);
        }
        try {
            if (!runner) {
                runner = new Function('data',code);
            }
            var run_data = {};
            try {
                 run_data = JSON.parse(line);
            } catch(err) {
            }
            result = runner(run_data);
            if (Array.isArray(result)) {
                console.log(JSON.stringify(result));
            } else {
                console.log('[]');
            }
        } catch(err) {
            console.log(JSON.stringify(['error',err.message,[err.stack]]));
            process.exit(0);
        }
    }
});

File: runner/runner.rb
#!/usr/bin/env ruby

require 'timeout'
require 'json'

type = :code
code_installed = false
code = <<CODE
def code_runner(data)
CODE

while data = gets
  case type
  when :code then
    if data.start_with?('__END__CODE__')
      code << "end"
      type = :run
      puts "READY"
      STDOUT.flush
      #STDERR.puts code.inspect
    else
      code << data
    end
  when :run then
    if data.start_with?('EXIT')
      STDOUT.flush
      exit
    end
    run_data = JSON.parse(data, :symbolize_names => true) rescue {}
    begin
      if !code_installed
        eval(code)
        code_installed = true
      end

      result = code_runner(run_data)
      if result.is_a?(Array)
        puts result.to_json
      else
        puts [].to_json
      end
    rescue Exception => e
      puts(['error',e.message,e.backtrace].to_json)
      STDOUT.flush
      exit
    end
    STDOUT.flush
    #STDERR.puts result.inspect
  end
end
